app/compare/diff_chunk.tsx

```diff
@@ -1,61 +1,113 @@
+import DiffMatchPatch from "diff-match-patch";
 import React from "react";
-import JsDiff from "diff";
 import { OutlinedButton } from "../components/button/button";
 
 export type DiffChunkProps = {
-  change: JsDiff.Change;
+  change: DiffMatchPatch.Diff;
   defaultExpanded: boolean;
 };
 
-type DiffChunkState = {
+type State = {
   expanded: boolean;
 };
 
-export default class DiffChunk extends React.Component<DiffChunkProps, DiffChunkState> {
-  state = { expanded: this.props.defaultExpanded };
+const MIN_COLLAPSED_UNCHANGED_REGION_SIZE = 16;
+const MIN_COLLAPSED_CHANGED_REGION_SIZE = 128;
+// Number of lines to show before and after a long segment of identical lines.
+const NUM_LINES_OF_CONTEXT = 4;
+
+export default class DiffChunk extends React.Component<DiffChunkProps, State> {
+  state: State = this.getInitialState();
+
+  private getInitialState() {
+    const [diffType] = this.props.change;
+    return { expanded: diffType === DiffMatchPatch.DIFF_EQUAL && this.props.defaultExpanded };
+  }
 
   componentDidUpdate(prevProps: DiffChunkProps) {
     if (prevProps.defaultExpanded !== this.props.defaultExpanded) {
-      this.setState({ expanded: this.props.defaultExpanded });
+      this.setState(this.getInitialState());
     }
   }
 
   private onExpand() {
     this.setState({ expanded: true });
   }
 
-  render() {
-    const {
-      change: { value, added, removed },
-    } = this.props;
-
-    const isUnchanged = !added && !removed;
-
-    const lines = value.trimEnd().split("\n");
-
-    const nodes = lines.map((line: string, index: number) => (
-      <div key={`${index}`} className={`diff-line ${added ? "added" : ""} ${removed ? "removed" : ""}`}>
-        <div className="plus-minus-cell">
-          {added && "+"}
-          {removed && "-"}
-        </div>
-        <div className="diff-line-content">{line}</div>
-      </div>
-    ));
-
-    if (nodes.length >= 10 && isUnchanged && !this.state.expanded) {
-      return [
-        nodes[0],
-        <OutlinedButton className="diff-line collapsed" onClick={this.onExpand.bind(this)}>
-          <div className="plus-minus-cell">
+  private renderAdded(text: string) {
+    return <ins className="added">{text}</ins>;
+  }
+
+  private renderRemoved(text: string) {
+    return (
+      <del className="removed">
+        <span>{text}</span>
+      </del>
+    );
+  }
+
+  private renderUnchanged(text: string) {
+    return <>{text}</>;
+  }
+
+  private renderChunk(
+    renderText: (text: string) => React.ReactNode,
+    collapsedLabel: string,
```

Comment:
are collapsedLabel and expandButtonClass ever different?

Reply:
They're effectively different data types (one is a class name and one is a UI message) so I figured they should be separate params. I didn't want someone to change the label and not realize that it's a class name, and then they change it to something that's an invalid as a class name. Also if we translate the UI at some point in the future it's a bug waiting to happen if these aren't separate.